runtime/bindings/python/hal.cc

```diff
@@ -455,6 +455,12 @@ py::object MapElementTypeToDType(iree_hal_element_type_t element_type) {
     case IREE_HAL_ELEMENT_TYPE_FLOAT_64:
       dtype_code = "d";
       break;
+    case IREE_HAL_ELEMENT_TYPE_COMPLEX_FLOAT_64:
+      dtype_code = "complex64";
+      break;
+    case IREE_HAL_ELEMENT_TYPE_COMPLEX_FLOAT_128:
+      dtype_code = "complex128";
+      break;
```

Comment:
Do you have a source for these codes? The reference linked above the switch statement (https://docs.python.org/3/c-api/arg.html#numbers) only lists

> D ([complex](https://docs.python.org/3/library/functions.html#complex)) [Py_complex *]
Convert a C [Py_complex](https://docs.python.org/3/c-api/complex.html#c.Py_complex) structure to a Python complex number.

I see a few other pages like https://numpy.org/doc/stable/user/basics.types.html#array-types-and-conversions-between-types but they don't quite look like what goes here (generally single character codes).

Reply:
Ah, https://numpy.org/doc/stable/reference/arrays.dtypes.html
![image](https://user-images.githubusercontent.com/4010439/229200723-7824ed86-cfcc-4b2c-b035-372e0a3c6297.png)

```python
for key in numpy.sctypeDict.keys():
  if "complex" in str(key):
    print(key)

# complex64
# complex128
# complex256
# complex_
# singlecomplex
# longcomplex
# complex
```

```python
dt = numpy.dtype("complex64")
print(dt.type)
print(dt.kind)
print(dt.char)
print(dt.num)

dt = numpy.dtype("complex128")
print(dt.type)
print(dt.kind)
print(dt.char)
print(dt.num)

# <class 'numpy.complex64'>
# c
# F
# 14
# <class 'numpy.complex128'>
# c
# D
# 15
```

Can you add that link and change the TODO text a bit near the top of this switch statement?
```
  // See: https://docs.python.org/3/c-api/arg.html#numbers
  // TODO: Handle dtypes that do not map to a code (i.e. fp16).
```

I'm also wondering what sort of test coverage we could add for this (probably some tests that use complex numbers in programs, run via the python APIs?)

Reply:
I'll make those changes and work on some test cases

Reply:
Ahh, good timing with this, I'm tracking down an issue on Windows where that function was buggy (due to how platforms implement integer types). I'm putting together a PR that makes some related changes (like adding that reference link and swapping "i" -> "int32")